from pathlib import Path
import pandas as pd

# Read configuration
configfile: "config/config.yaml"
f5 = Path(config["f5"])
fq = Path(config["fq"])
SOURCEDIR = Path(config["sourcedir"])
OUTDIR = Path(config["results"])


SAMPLE_DATA = pd.read_csv(config["sample_file"], comment="#")
SAMPLES = SAMPLE_DATA["samplename"]

if len(SAMPLES) != len(set(SAMPLES)):
    raise RuntimeError("\n*** Samplenames are NOT unique. Please fix your sample definition file. Cannot continue... exiting. ***\n")

# SAMPLE_LOCATION is a dict that can be used to find the .fastq files that belong to a particular sample
SAMPLE_LOCATION = {i["samplename"] : f"{SOURCEDIR}/{i['samplename']}.fastq"
    for idx, i in SAMPLE_DATA.iterrows()}



include: "rules/nanopolish.smk"

rule all:
    input:
        #index = expand(fq / "{sample}.fastq.index", sample=samples),
        #fai = expand(fq / "{sample}.fastq.index.fai", sample=samples),
        #gzi = expand(fq /"{sample}.fastq.index.gzi", sample=samples),
        #readdb = expand(fq / "{sample}.fastq.index.readdb", sample=samples),
        #sorted= expand("results/minimap2_results/{sample}.bam", sample=samples),
        #meth = expand("results/nanopolish_results/{sample}_methylation.tsv", sample=SAMPLES),
        freq = expand("results/nanopolish_results/{sample}_frequency.tsv", sample=SAMPLES)

